### PR TITLE
improving _core_computation

### PR DESCRIPTION
**Pull request by adamwlev**
Hello,

Was playing around with this last night and ran into memory issues. I was trying to get some confidence estimates for a forest with 1 million training examples and 300k testing examples. The line `cov_hat = np.zeros((X_train.shape[0], X_test.shape[0]))` was giving me Memory Errors since having this matrix in memory would take over 300GB.

I realized a couple things:

1. The function _core_computation was not fully vectorized. It was implemented by summing outer products instead of just one matrix multiplication. With my testing, doing this multiplication all at once is around 5-10x faster with no additional memory requirements.

2. If a matrix of shape (n_train_sample,n_test_sample) cannot fit in memory, this function did not work at all. However, this is easy to deal with since we can deal with chunks of the pred_centered matrix at a time. I implemented this by adding a couple arguments to _core_computation and random_forest_error that allow the user to specify if they have memory constraints and if so what is the maximum memory an intermediate matrix should take up in megabytes. To my surprise, if memory_limit is tuned decently, this method is actually even faster than the whole matrix at one time method by around 3x (even if it's unnecessary and the whole matrix can fit in memory).

times_for_7k_by_9k

In summary, these improvements can speedup the _core_computation function up to 15-30x if the user plays with the memory_limit arguments or 5-10x if he or she does not. It also allows the function to work when a matrix of shape (n_train_sample,n_test_sample) cannot fit into memory.

All tests currently in the repo pass on python 2 and 3. I can add tests for when memory_constrained=True if it is desired. Please let me know if you have any comments or questions.

Thanks

### CODE REVIEW

File: forestci/forestci.py [hunk 1]
```diff
@@ -57,14 +57,23 @@ def calc_inbag(n_samples, forest):
...
 
-def _core_computation(X_train, X_test, inbag, pred_centered, n_trees):
-    cov_hat = np.zeros((X_train.shape[0], X_test.shape[0]))
-
-    for t_idx in range(n_trees):
-        inbag_r = (inbag[:, t_idx] - 1).reshape(-1, 1)
-        pred_c_r = pred_centered.T[t_idx].reshape(1, -1)
-        cov_hat += np.dot(inbag_r, pred_c_r) / n_trees
-    V_IJ = np.sum(cov_hat ** 2, 0)
+def _core_computation(X_train, X_test, inbag, pred_centered, n_trees, memory_constrained=False, memory_limit=None):
```
Comment (arokem) [inline forestci/forestci.py]: Please wrap this line to 80 characters (PEP8)

File: forestci/forestci.py [hunk 2]
```diff
@@ -78,7 +87,7 @@ def _bias_correction(V_IJ, inbag, pred_centered, n_trees):
     return V_IJ_unbiased
 
 
-def random_forest_error(forest, X_train, X_test, inbag=None):
+def random_forest_error(forest, X_train, X_test, inbag=None, memory_constrained=False, memory_limit=None):
```
Comment (arokem) [inline forestci/forestci.py]: Please wrap to 80 characters

File: forestci/forestci.py [hunk 3]
```diff
@@ -125,7 +147,7 @@ def random_forest_error(forest, X_train, X_test, inbag=None):
     pred_mean = np.mean(pred, 0)
     pred_centered = pred - pred_mean
     n_trees = forest.n_estimators
-    V_IJ = _core_computation(X_train, X_test, inbag, pred_centered, n_trees)
+    V_IJ = _core_computation(X_train, X_test, inbag, pred_centered, n_trees, memory_constrained, memory_limit)
```
Comment (arokem) [inline forestci/forestci.py]: Please wrap to 80 characters (PEP8)

**Review by arokem — CHANGES_REQUESTED** [review]
Thank you for this contribution and apologies for how long it's taken us to get to this! 
Before we can merge this, I would like to ask that you add some tests that exercise the different modes of the core computation. That is, we'd like to make sure that the results come out the same. To do so, you might add a few lines in https://github.com/scikit-learn-contrib/forest-confidence-interval/blob/master/forestci/tests/test_forestci.py

**Comment by adamwlev** [comment]
thanks @arokem. Added some lines to the test_core_computation function. Had to repeat test examples 1000 times so that matrix would be big enough to meaningfully use the memory_constrained functionality. Also fixed for PEP8. All tests seems to be passing. Thoughts?

**Review by arokem — APPROVED** [review]

**Comment by arokem** [comment]
LGTM - merging. Thanks!
